lib/charms/tls_certificates_interface/v1/tls_certificates.py

```diff
@@ -0,0 +1,850 @@
+        ]
+
+
+class TLSCertificatesRequires(Object):
+    """TLS certificates requirer class to be instantiated by TLS certificates requirers."""
+
+    on = CertificatesRequirerCharmEvents()
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relationship_name: str,
+        expiry_notification_time: int = 168,
+    ):
+        """Generates/use private key and observes relation changed event.
+
+        Args:
+            charm: Charm object
+            relationship_name: Juju relation name
+            expiry_notification_time (int): Time difference between now and expiry. Used to
+                trigger the CertificateExpiring event.
+        """
+        super().__init__(charm, relationship_name)
+        self.relationship_name = relationship_name
+        self.charm = charm
+        self.expiry_notification_time = expiry_notification_time
+        self.framework.observe(
+            charm.on[relationship_name].relation_changed, self._on_relation_changed
+        )
+        self.framework.observe(charm.on.update_status, self._on_update_status)
+
+    def request_certificate(self, csr: bytes) -> None:
+        """Request TLS certificate to provider charm.
+
+        Args:
+            csr (bytes): Certificate Signing Request
+
+        Returns:
+            None
+        """
+        logger.info("Received request to create certificate")
+        relation = self.model.get_relation(self.relationship_name)
+        if not relation:
+            message = (
+                f"Relation {self.relationship_name} does not exist - "
+                f"The certificate request can't be completed"
+            )
+            logger.error(message)
+            raise RuntimeError(message)
+        relation_data = load_unit_relation_data(relation.data[self.model.unit])
+        new_certificate_request = {"certificate_signing_request": csr.decode()}
+        if "certificate_signing_requests" in relation_data:
+            certificate_request_list = relation_data["certificate_signing_requests"]
+            if new_certificate_request in certificate_request_list:
+                logger.info("Request was already made - Doing nothing")
+                return
+            certificate_request_list.append(new_certificate_request)
+        else:
+            certificate_request_list = [new_certificate_request]
+        relation.data[self.model.unit]["certificate_signing_requests"] = json.dumps(
+            certificate_request_list
+        )
+        logger.info("Certificate request sent to provider")
+
+    def renew_certificate(self, certificate_signing_request: bytes) -> None:
+        """Renews Certificate.
+
+        Revokes certificate than requests one with the same CSR.
+
+        Args:
+            certificate_signing_request:
+
+        Returns:
+            None
+        """
+        self.revoke_certificate(certificate_signing_request)
+        self.request_certificate(certificate_signing_request)
```

Comment:
This looks like an outage waiting to happen to me. If the `request_certificate` call fails for some reason, our certificate is already revoked.

Reply:
Yeah I get the risk but in the context of renewal, the new csr will be the same as the old one. If we just request a new certificate for it, the relation data won't change. We need to send information to the provider to both revoke the old cert and create a new one.

Reply:
Ok, and the docstring mentions this behaviour, so charmers can take it into account.